src/client/opamClient.ml

```diff
@@ -1962,62 +1957,109 @@ let check_installed ~build ~post t atoms =
+                        in
+                        match installed_conj with
+                        | Some conj ->
+                          get_found_and_missing (conj :: found_conj)
+                            missing_conj xs
+                        | None ->
+                          (* TODO: maybe it would be nice to display
+                             disjunctions correctly in the output instead of
+                             transforming everything into one big conjunction *)
+                          get_found_and_missing found_conj
+                            (disj @ missing_conj) xs
+                    in
+                    let (found_conj, missing_conj) =
+                      get_found_and_missing [] [] cnf_formula
+                    in
+                    (missing_conj = [], Some (missing_conj, found_conj)))
+                versions (false, None)
+          in
+          match missing_opt with
+          | None ->
+            (* No version *)
+            OpamPackage.Name.Map.add pkgname
+              (OpamPackage.Name.Set.singleton pkgname)
+              map,
+            seen
+          | Some (missing_conj, found_conj) ->
+            let missing_set =
+              List.fold_left (fun names (name, _) ->
+                  OpamPackage.Name.Set.add name names)
+                OpamPackage.Name.Set.empty missing_conj
             in
-            let missing_conj =
-              List.filter
-                (List.for_all (fun ((n,_vc) as atom) ->
-                     OpamPackage.Set.for_all
-                       (fun p -> not (OpamFormula.check atom p))
-                       (OpamPackage.packages_of_name t.installed n)))
-                cnf_formula
+            let new_map =
+              if OpamPackage.Name.Set.is_empty missing_set
+              then map
+              else OpamPackage.Name.Map.add pkgname missing_set map
             in
-            if missing_conj = [] then true, None
-            else false, Some (pkg,missing_conj))
-          versions (false,None)
-      in
-      if compliant then map else
-      match missing_opt with
-      | None -> assert false (* version set can't be empty *)
-      | Some (pkg, missing_conj) ->
-        OpamPackage.Map.add pkg
-          (List.fold_left (fun names disj ->
-               List.fold_left (fun names (name, _) ->
-                   OpamPackage.Name.Set.add name names)
-                 names disj)
-              OpamPackage.Name.Set.empty missing_conj)
-          map
-    ) pkgs OpamPackage.Map.empty
+            if found_conj = [] || not recursive
+            then (new_map, seen)
+            else loop new_map seen found_conj
+      ) (map, seen) atoms
+  in
+  fst (loop OpamPackage.Name.Map.empty SeenSet.empty atoms)
 
 let assume_built_restrictions ?available_packages t atoms =
-  let missing = check_installed ~build:false ~post:false t atoms in
+  let missing =
+    check_installed ~build:false ~post:false ~recursive:false t atoms
```

Comment:
Assume built does not check recursively, is it to not change the known behaviour?

Reply:
no, the previous behaviour of this function did not check recursively so the behaviour of `assume-built` hasn't changed either